src/model.jl

```diff
@@ -104,100 +98,179 @@ implementation simply uses `T.name.wrapper` which is the `UnionAll` type corresp
-        erorr("cannot set :fieldname index")
-    else
-        newparent = if nm in keys(m)
-            _setindex(parent(m), Tuple(x), nm)
-        else                                 
-            _addindex(parent(m), Tuple(x), nm)
-        end
-        setparent!(m, newparent)
+_keys(params::Tuple, ::AbstractModel) = (:component, :fieldname, keys(first(params))...)
+_keys(::Tuple{}, ::AbstractModel) = ()
+_isreserved(key::Symbol) = key == :component || key == :fieldname
+@inline _enumerate(tup::NTuple{N,Any}) where N = map(tuple, tuple(1:N...), tup) # type stable tuple enumeration
+
+# Indexing kernels
+@inline _getindex(ps::Tuple{Vararg{<:Param}}, i) = _getindex(ps, i, :)
+@inline _getindex(ps::Tuple{Vararg{<:Param}}, col::Symbol) = _getindex(ps, :, col)
+@inline _getindex(ps::Tuple{Vararg{<:Param}}, i, ::Colon) = ps[i]
+@inline _getindex(ps::Tuple{Vararg{<:Param}}, i, col::Symbol) = map(p -> p[col], ps[i])
+@inline _getindex(ps::Tuple{Vararg{<:Param}}, i::AbstractVector{Bool}, col::Symbol) = _getindex(ps, findall(i), col)
+@inline _setindex(obj, xs, ::Colon, cols) = _setindex(obj, xs, 1:length(params(obj)), cols)
+@inline _setindex(obj, xs, idxs::AbstractVector{Bool}, cols::AbstractVector) = _setindex(obj, xs, findall(idxs), cols)
+@inline _setindex(obj, xs, idxs::AbstractVector{Bool}, ::Type{Val{col}}) where col = _setindex(obj, xs, findall(idxs), Val{col})
+@inline @generated _setindex(ps::Tuple{Vararg{<:Param}}, x, i::Integer, ::Type{Val{col}}) where col = :(@set ps[i].$col = x)
+@inline function _setindex(obj, x, i::Integer, ::Type{Val{col}}) where col
+    ps = params(obj)
+    newps = _setindex(ps, x, i, Val{col})
+    return Flatten.reconstruct(obj, newps, SELECT, IGNORE)
+end
+@inline function _setindex(obj, xs, ::Colon, ::Type{Val{col}}) where col
+    # handle special case for ::Colon (all indices) where we can be type stable
+    ps = params(obj)
+    newps = map(_enumerate(ps)) do (i,p)
+        _setindex((p,), xs[i], 1, Val{col})[1]
     end
+    return Flatten.reconstruct(obj, newps, SELECT, IGNORE)
 end
-# TODO do this with lenses
-@inline function _setindex(obj, xs::Tuple, nm::Symbol)
-    lens = Setfield.PropertyLens{nm}()
-    newparams = map(params(obj), xs) do par, x
-        Param(Setfield.set(parent(par), lens, x))
+@inline function _setindex(obj, xs, idxs::AbstractVector, ::Type{Val{col}}) where col
+    ps = params(obj)
+    for i in 1:length(idxs)
+        ps = _setindex(ps, xs[i], idxs[i], Val{col})
     end
-    Flatten.reconstruct(obj, newparams, SELECT, IGNORE)
+    return Flatten.reconstruct(obj, ps, SELECT, IGNORE)
 end
-@inline function _addindex(obj, xs::Tuple, nm::Symbol)
-    newparams = map(params(obj), xs) do par, x
-        Param((; parent(par)..., (nm => x,)...))
+@inline function _addindex(obj, xs, ::Type{Val{col}}) where col
+    newparams = map(params(obj), xs) do p, x
+        Param((; parent(p)..., col => x))
     end
     Flatten.reconstruct(obj, newparams, SELECT, IGNORE)
```

Comment:
```suggestion
    return Flatten.reconstruct(obj, newparams, SELECT, IGNORE)
```

May as well make this change here too